spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -405,14 +420,35 @@ public Scan build() {
 
   private Scan buildBatchScan() {
     Schema expectedSchema = schemaWithMetadataColumns();
+    org.apache.iceberg.Scan scan =
+        buildIcebergBatchScan(false /* not include Column Stats */, expectedSchema);
+    if (pushedLimit != null && pushedLimit > 0 && hasDeletes(scan)) {
+      LOG.info("Skipping limit pushdown: detected row level deletes");
+      pushedLimit = null;
+    }
     return new SparkBatchQueryScan(
         spark,
         table,
-        buildIcebergBatchScan(false /* not include Column Stats */, expectedSchema),
+        scan,
         readConf,
         expectedSchema,
         filterExpressions,
-        metricsReporter::scanReport);
+        metricsReporter::scanReport,
+        pushedLimit);
+  }
+
+  private boolean hasDeletes(org.apache.iceberg.Scan scan) {
```

Comment:
This is too expensive. We should be using snapshot summary.

Reply:
That's optional in the spec, and delete content is also optional even if the snapshot summary is present. Can we rely on that here? 

I think we would have to fall back to the scan anyway if they are missing

Reply:
In a few places, we default to what's safe. In this case it would be assuming there are deletes. We can also check if there are delete manifests in the snapshot we read, but doing an extra scan and iterating through all files seems too expensive.